Routes are misleading

I tried to setup Open-VSX. I found #703 very helpful to do so using docker-compose.
However now I'm facing the issue that every route is leading to the main page (checked on clients and the server itself) which results in the following error:

![image](https://github.com/eclipse/openvsx/assets/11911587/a3f7752a-59ec-4c6a-b758-21dc2c7276f2)

Is there any way to fix it? Most probably a misconfiguration on my side.

application.yml
```
server:
  address: 0.0.0.0
  port: 8080
  shutdown: graceful
  compression:
    enabled: true
    mime-types: text/html,text/plain,text/css,application/javascript,application/json,text/xml,application/xml,application/xml+rss,text/javascript
    min-response-size: 1024
spring:
  cache:
    jcache:
      config: classpath:ehcache.xml
  datasource:
    url: jdbc:postgresql://postgres:5432/
    username: postgres
    password: postgres
    hikari:
      maximum-pool-size: 10
      leak-detection-threshold: 30000
  flyway:
    baseline-on-migrate: false
  jpa:
    open-in-view: false
    properties:
      hibernate:
        dialect: org.hibernate.dialect.PostgreSQLDialect
  resources:
    cache:
      cachecontrol:
        max-age: 4h
        cache-public: true
  session:
    store-type: jdbc
  lifecycle:
    timeout-per-shutdown-phase: 10s
  security:
    oauth2:
      client:
        registration:
          github:
            client-id: dummy
            client-secret: dummy
management:
  server:
    port: 8081
  health:
    probes:
      enabled: true
org:
  jobrunr:
    job-scheduler:
      enabled: true
    background-job-server:
      enabled: true
      worker-count: 2
    dashboard:
      enabled: false
    database:
      type: sql
    miscellaneous:
      allow-anonymous-data-usage: false
bucket4j:
  enabled: false
ovsx:
  webui:
    frontendRoutes: "/extension/**,/namespace/**,/user-settings/**,/admin-dashboard/**,/about,/publisher-agreement-*,/terms-of-use"
  elasticsearch:
    host: elasticsearch:9200
    enabled: true
    ssl: false
    relevance:
      rating: 0.2
      downloads: 1.0
      timestamp: 3.0
```

docker-compose.yml
```
version: "3.9"
services:
  openvsx-webui:
    image: ghcr.io/eclipse/openvsx-webui:v0.9.5
    ports:
      - "3000:3000"
    depends_on:
      openvsx-server:
        condition: service_healthy
  openvsx-server:
    image: ghcr.io/eclipse/openvsx-server:v0.9.5
    ports:
      - "8080:8080"
    healthcheck:
      test: ["CMD-SHELL", "curl http://openvsx-server:8080"]
      interval: 10s
      timeout: 10s
      retries: 6
    volumes:
      - ./application.yml:/home/openvsx/server/config/application.yml
    depends_on:
      elasticsearch:
        condition: service_healthy
      postgres:
        condition: service_healthy
  elasticsearch:
    image: elasticsearch:7.11.0
    environment:
      - xpack.security.enabled=false
      - discovery.type=single-node
    ports:
      - "9200:9200"
      - "9300:9300"
    healthcheck:
      test: ["CMD-SHELL", "curl http://elasticsearch:9200/_cluster/health?wait_for_status=green"]
      timeout: 60s
  postgres:
    image: postgres:10.17-alpine
    environment:
      - POSTGRES_PASSWORD=postgres
    ports:
      - "5432:5432"
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 10s
      timeout: 10s
      retries: 6
networks:
  default:
    name: openvsx-network
```

Also another Extension (RapidAPI.vscode-rapidapi-client) which has no version published yet.
I published the dependency (RapidAPI.vscode-services) first and when trying to publish RapidAPI.vscode-rapidapi-client the following error happens.

It created one entry in extension table and one in extension_version table. However no file_resource entries for the extension_version reference.

```
2023-06-01 11:08:29.248  WARN 1 --- [         task-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 23505
2023-06-01 11:08:29.248 ERROR 1 --- [         task-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : ERROR: duplicate key value violates unique constraint "unique_extension_version"
  Detail: Key (extension_id, target_platform, version)=(23917, universal, 1.10.2) already exists.
2023-06-01 11:08:29.249  INFO 1 --- [         task-1] o.h.e.j.b.internal.AbstractBatchImpl     : HHH000010: On release of batch it still contained JDBC statements
2023-06-01 11:08:30.270  WARN 1 --- [         task-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 23505
2023-06-01 11:08:30.270 ERROR 1 --- [         task-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : ERROR: duplicate key value violates unique constraint "unique_extension_version"
  Detail: Key (extension_id, target_platform, version)=(23917, universal, 1.10.2) already exists.
2023-06-01 11:08:30.270  INFO 1 --- [         task-1] o.h.e.j.b.internal.AbstractBatchImpl     : HHH000010: On release of batch it still contained JDBC statements
2023-06-01 11:08:31.290  WARN 1 --- [         task-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 23505
2023-06-01 11:08:31.290 ERROR 1 --- [         task-1] o.h.engine.jdbc.spi.SqlExceptionHelper   : ERROR: duplicate key value violates unique constraint "unique_extension_version"
  Detail: Key (extension_id, target_platform, version)=(23917, universal, 1.10.2) already exists.
2023-06-01 11:08:31.290  INFO 1 --- [         task-1] o.h.e.j.b.internal.AbstractBatchImpl     : HHH000010: On release of batch it still contained JDBC statements
2023-06-01 11:08:31.297 ERROR 1 --- [         task-1] .a.i.SimpleAsyncUncaughtExceptionHandler : Unexpected exception occurred invoking async method: public void org.eclipse.openvsx.publish.PublishExtensionVersionHandler.publishAsync(org.eclipse.openvsx.entities.FileResource,org.eclipse.openvsx.util.TempFile,org.eclipse.openvsx.ExtensionService)

org.springframework.dao.DataIntegrityViolationException: could not execute statement; SQL [n/a]; constraint [unique_extension_version]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:276) ~[spring-orm-5.3.21.jar:5.3.21]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:233) ~[spring-orm-5.3.21.jar:5.3.21]
        at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:566) ~[spring-orm-5.3.21.jar:5.3.21]
        at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:743) ~[spring-tx-5.3.21.jar:5.3.21]
        at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:711) ~[spring-tx-5.3.21.jar:5.3.21]
        at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:654) ~[spring-tx-5.3.21.jar:5.3.21]
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:407) ~[spring-tx-5.3.21.jar:5.3.21]
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:163) ~[spring-retry-1.3.3.jar:na]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.eclipse.openvsx.publish.PublishExtensionVersionService$$EnhancerBySpringCGLIB$$fe9b423f.persistResource(<generated>) ~[classes/:na]
        at org.eclipse.openvsx.publish.PublishExtensionVersionHandler.publishAsync(PublishExtensionVersionHandler.java:214) ~[classes/:na]
        at org.eclipse.openvsx.publish.PublishExtensionVersionHandler$$FastClassBySpringCGLIB$$8d4b46cf.invoke(<generated>) ~[classes/:na]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:97) ~[spring-retry-1.3.3.jar:na]
        at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329) ~[spring-retry-1.3.3.jar:na]
        at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209) ~[spring-retry-1.3.3.jar:na]
        at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:133) ~[spring-retry-1.3.3.jar:na]
        at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:160) ~[spring-retry-1.3.3.jar:na]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) ~[spring-aop-5.3.21.jar:5.3.21]
        at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:115) ~[spring-aop-5.3.21.jar:5.3.21]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
        at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:109) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3375) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3908) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684) ~[na:na]
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1407) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:489) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3290) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2425) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:449) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:183) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$300(JdbcResourceLocalTransactionCoordinatorImpl.java:40) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:281) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:101) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:562) ~[spring-orm-5.3.21.jar:5.3.21]
        ... 30 common frames omitted
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "unique_extension_version"
  Detail: Key (extension_id, target_platform, version)=(23917, universal, 1.10.2) already exists.
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675) ~[postgresql-42.3.6.jar:42.3.6]
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365) ~[postgresql-42.3.6.jar:42.3.6]
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355) ~[postgresql-42.3.6.jar:42.3.6]
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490) ~[postgresql-42.3.6.jar:42.3.6]
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408) ~[postgresql-42.3.6.jar:42.3.6]
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:167) ~[postgresql-42.3.6.jar:42.3.6]
        at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:135) ~[postgresql-42.3.6.jar:42.3.6]
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61) ~[HikariCP-4.0.3.jar:na]
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java) ~[HikariCP-4.0.3.jar:na]
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
        ... 51 common frames omitted
```

--- a/server/build.gradle
+++ b/server/build.gradle
@@ -23,7 +23,7 @@ def versions = [
     springdoc: '2.1.0',
     spdx: '2.2.8',
     gcloud: '2.22.3',
-    azure: '12.18.0',
+    azure: '12.23.0',
     guava: '30.0-jre',
     junit: '5.7.1',
     testcontainers: '1.15.2',

--- a/server/src/main/java/org/eclipse/openvsx/LocalRegistryService.java
+++ b/server/src/main/java/org/eclipse/openvsx/LocalRegistryService.java
@@ -113,7 +113,7 @@ public class LocalRegistryService implements IExtensionRegistry {
     @Cacheable(value = CACHE_EXTENSION_JSON, keyGenerator = GENERATOR_EXTENSION_JSON)
     public ExtensionJson getExtension(String namespace, String extensionName, String targetPlatform, String version) {
         var extVersion = findExtensionVersion(namespace, extensionName, targetPlatform, version);
-        var json = toExtensionVersionJson(extVersion, targetPlatform, true, false);
+        var json = toExtensionVersionJson(extVersion, targetPlatform, true);
         json.downloads = getDownloads(extVersion.getExtension(), targetPlatform, extVersion.getVersion());
         return json;
     }
@@ -611,24 +611,13 @@ public class LocalRegistryService implements IExtensionRegistry {
         return ResultJson.success("Valid token");
     }
 
-    @Retryable(value = { DataIntegrityViolationException.class })
-    @Transactional(rollbackOn = ErrorResultException.class)
     public ExtensionJson publish(InputStream content, UserData user) throws ErrorResultException {
-        var token = new PersonalAccessToken();
-        token.setDescription("One time use publish token");
-        token.setValue(users.generateTokenValue());
-        token.setCreatedTimestamp(TimeUtil.getCurrentUTC());
-        token.setAccessedTimestamp(token.getCreatedTimestamp());
-        token.setUser(user);
-        token.setActive(true);
-        entityManager.persist(token);
-
-        var json = publish(content, token.getValue());
-        token.setActive(false);
+        var token = users.createAccessToken(user, "One time use publish token");
+        var json = publish(content, token.value);
+        users.deleteAccessToken(user, token.id);
         return json;
     }
 
-    @Transactional(rollbackOn = ErrorResultException.class)
     public ExtensionJson publish(InputStream content, String tokenValue) throws ErrorResultException {
         var token = users.useAccessToken(tokenValue);
         if (token == null || token.getUser() == null) {
@@ -639,7 +628,7 @@ public class LocalRegistryService implements IExtensionRegistry {
         eclipse.checkPublisherAgreement(token.getUser());
 
         var extVersion = extensions.publishVersion(content, token);
-        var json = toExtensionVersionJson(extVersion, null, true, true);
+        var json = toExtensionVersionJson(extVersion, null, true);
         json.success = "It can take a couple minutes before the extension version is available";
 
         var sameVersions = repositories.findVersions(extVersion.getVersion(), extVersion.getExtension());
@@ -812,14 +801,10 @@ public class LocalRegistryService implements IExtensionRegistry {
         }).collect(Collectors.toList());
     }
 
-    public ExtensionJson toExtensionVersionJson(ExtensionVersion extVersion, String targetPlatform, boolean onlyActive, boolean inTransaction) {
+    public ExtensionJson toExtensionVersionJson(ExtensionVersion extVersion, String targetPlatform, boolean onlyActive) {
         var extension = extVersion.getExtension();
-        var latest = inTransaction
-                ? versions.getLatest(extension, targetPlatform, false, onlyActive)
-                : versions.getLatestTrxn(extension, targetPlatform, false, onlyActive);
-        var latestPreRelease = inTransaction
-                ? versions.getLatest(extension, targetPlatform, true, onlyActive)
-                : versions.getLatestTrxn(extension, targetPlatform, true, onlyActive);
+        var latest = versions.getLatestTrxn(extension, targetPlatform, false, onlyActive);
+        var latestPreRelease = versions.getLatestTrxn(extension, targetPlatform, true, onlyActive);
 
         var json = extVersion.toExtensionJson();
         json.preview = latest != null && latest.isPreview();

--- a/server/src/main/java/org/eclipse/openvsx/admin/AdminAPI.java
+++ b/server/src/main/java/org/eclipse/openvsx/admin/AdminAPI.java
@@ -199,7 +199,7 @@ public class AdminAPI {
                 json.allVersions = Collections.emptyMap();
                 json.allTargetPlatformVersions = Collections.emptyMap();
             } else {
-                json = local.toExtensionVersionJson(latest, null, false, false);
+                json = local.toExtensionVersionJson(latest, null, false);
                 json.allTargetPlatformVersions = versions.getVersionsTrxn(extension).stream()
                         .collect(Collectors.groupingBy(ExtensionVersion::getVersion, Collectors.mapping(ExtensionVersion::getTargetPlatform, Collectors.toList())));
             }

--- a/server/src/main/java/org/eclipse/openvsx/publish/PublishExtensionVersionHandler.java
+++ b/server/src/main/java/org/eclipse/openvsx/publish/PublishExtensionVersionHandler.java
@@ -63,7 +63,7 @@ public class PublishExtensionVersionHandler {
     @Autowired
     ExtensionValidator validator;
 
-    @Transactional
+    @Transactional(rollbackOn = ErrorResultException.class)
     public ExtensionVersion createExtensionVersion(ExtensionProcessor processor, PersonalAccessToken token, LocalDateTime timestamp, boolean checkDependencies) {
         // Extract extension metadata from its manifest
         var extVersion = createExtensionVersion(processor, token.getUser(), token, timestamp);

--- a/server/src/main/java/org/eclipse/openvsx/publish/PublishExtensionVersionService.java
+++ b/server/src/main/java/org/eclipse/openvsx/publish/PublishExtensionVersionService.java
@@ -43,6 +43,7 @@ public class PublishExtensionVersionService {
         repositories.findFiles(extVersion).forEach(entityManager::remove);
     }
 
+    @Retryable
     public void storeDownload(FileResource download, TempFile extensionFile) {
         if (storageUtil.shouldStoreExternally(download)) {
             storageUtil.uploadFile(download, extensionFile);
@@ -78,8 +79,7 @@ public class PublishExtensionVersionService {
     }
 
     @Transactional
-    public void persistResource(FileResource resource) {
-        resource.setExtension(entityManager.merge(resource.getExtension()));
+    public synchronized void persistResource(FileResource resource) {
         entityManager.persist(resource);
     }
 

